MTKA-1265: Make relationship fields with back references editable from the reverse side

matthewguywright opened the pull request:
## Description
Allows the display of reverse relationships on the publisher entry screen.

matthewguywright committed:
working with model and base relationships

matthewguywright committed:
updated for empty fields

matthewguywright committed:
updated per PR

matthewguywright committed:
Update class-publisher-form-editing-experience.php

matthewguywright committed:
current commits

matthewguywright committed:
Merge branch 'main' into feature/mw-bh-1265-rev-rels

matthewguywright committed:
Update class-publisher-form-editing-experience.php

matthewguywright committed:
Merge branch 'main' into feature/mw-bh-1265-rev-rels

matthewguywright committed:
Update CreateRelationshipFieldEntryCest.php

matthewguywright committed:
working version

matthewguywright committed:
Update class-publisher-form-editing-experience.php

nickcernis committed:
chore: remove log statement

nickcernis committed:
test: check for reverse relationship fields

nickcernis commented:
@mindctrl @matthewguywright I completed tests above in 8abf508ec1b8035c527c401e33f1f75f81a37c71. These exclude tests for saving data, which needs to be fixed as part of MTKA-1291.

This is ready to review.

I suggest we do not merge to main until MTKA-1291 is also ready to merge (fixing reverse relationship fields so they save data). That way we keep main in a deployable state. I'm working on 1291 now.

mindctrl commented on tests/acceptance/CreateRelationshipFieldEntryCest.php:
```diff
@@ -3,35 +3,14 @@
 	}
 
+	/**
+	 * @before create_company_employee_models
```
Never noticed this feature before... nice 😎

matthewguywright committed:
Merge branch 'main' into feature/mw-bh-1265-rev-rels

nickcernis committed:
[MTKA 1291] Enable saving relationships from the reverse side (#348)

* refactor: extract reverse relations append logic

This needs to be used elsewhere when searching
field groups by slug.

It also makes it more testable independently.

* refactor: adjust field lookup logic

To account for the new function signature that will
also get reverse relationship fields from other models.

* test: adjust tests to use new signature

Also adds a test for reverse references.

* refactor: remove unused import

Left over from removal of relationship fields?

There is no longer a function named get_top_level_fields.

* test: add unit tests for appending reverse fields

* test: can update related entries on reverse side

nickcernis commented:
The PHP unit test fail after the recent merge here seems to relate to WP requiring a new third-party dependency to run unit tests under PHPUnit 8.

Error:

```
> phpunit
Error: The PHPUnit Polyfills library is a requirement for running the WP test suite.
If you are trying to run plugin/theme integration tests, make sure the PHPUnit Polyfills library (https://github.com/Yoast/PHPUnit-Polyfills) is available and either load the autoload file of this library in your own test bootstrap before calling the WP Core test bootstrap file; or set the absolute path to the PHPUnit Polyfills library in a "WP_TESTS_PHPUNIT_POLYFILLS_PATH" constant to allow the WP Core bootstrap to load the Polyfills.
```

References:
- https://make.wordpress.org/core/2021/09/27/changes-to-the-wordpress-core-php-test-suite/ (includes a section titled “Step-by-step: how to make your test setup compatible with these changes and with higher PHPUnit versions”).
- https://core.trac.wordpress.org/ticket/53945
- https://github.com/Yoast/PHPUnit-Polyfills

This may also affect other branches when tests run there.

mindctrl committed:
Merge branch 'main' into feature/mw-bh-1265-rev-rels

mindctrl commented:
This PR relies on #353 and should not be merged until that one is merged into here.

nickcernis committed:
[MTKA-1319] Fix relationship cardinality restrictions on the reverse side (#353)

* fix: use singular name for many to one cardinality

So that the button reads “Link Right” instead of
“Link Rights” for many to one fields.

* fix: invert cardinality for reverse relationships

* Add PHPUnit Polyfills library (#354)

* chore: Add yoast/phpunit-polyfills composer dependency

This fixes test failure issues due to changes made in the
WordPress Core PHP test suite.

See https://make.wordpress.org/core/2021/09/27/changes-to-the-wordpress-core-php-test-suite/

* test: rename setUp and tearDown fixture methods in phpunit tests.

This ensure compatibility with the PHPUnit Polyfill library
and enables testing on multiple versions of PHPUnit.

* chore: remove unnecessary setUp method in test-relationships-db.php
and rename tearDown method.

* test: remove current version from database before
running UpdaterTestCases::test_update_plugin_no_saved_version.

* fix: add cardinality property to relationship test fields

* test: confirm reverse cardinality is enforced

* test: confirm cardinality is reversed

Co-authored-by: John Parris <john.parris@wpengine.com>

nickcernis commented:
This is now ready for a final review.

nickcernis committed:
docs: copy tweaks
